ddl/placement/bundle.go

```diff
@@ -363,16 +363,56 @@ func (b *Bundle) Tidy() error {
 }
 
 // Reset resets the bundle ID and keyrange of all rules.
-func (b *Bundle) Reset(newID int64) *Bundle {
-	b.ID = GroupID(newID)
-	// Involve all the table level objects.
-	startKey := hex.EncodeToString(codec.EncodeBytes(nil, tablecodec.GenTablePrefix(newID)))
-	endKey := hex.EncodeToString(codec.EncodeBytes(nil, tablecodec.GenTablePrefix(newID+1)))
-	for _, rule := range b.Rules {
-		rule.GroupID = b.ID
-		rule.StartKeyHex = startKey
-		rule.EndKeyHex = endKey
+func (b *Bundle) Reset(ruleIndex int, newIDs []int64) *Bundle {
```

Comment:
The current code about bundles seems some what weird and a little complete now. How about refactoring them to below methods:

```
// Build bundle from table info, and the partition rules and included in the return
NewBundleFromTable(tbl *model.TableInfo) (*Bundle, error)

// Build bundle from partition
NewBundleFromPartition(partition *model.PartitionDefinition) (*Bundle, error)
```

I think the above methods is enough for ddl to use. It's not necessary to use `BundleByName` in infoschema because we'll finally delete it and the logic that directly building bundle from meta is more brief and easy to maintain.

We do not need method `Tidy` too. We can `tidy` the bundle when we constructing it, I think no where requires a not tidied bundle.

Another advices is the reuse of the code, for example, we can introduce a private method like 

`newBundleFromPolicyOrDirectOptions(infoschema.InfoSchema, *model.PolicyRefInfo, *model.PlacementSettings) (*Bundle, error)`

It will return a semi-finished Bundle and just for code reuse and it should only be visible in placement package

Reply:
1: refactored to `func NewBundleFromTblInfo(t *meta.Meta, job *model.Job, tbInfo *model.TableInfo)` and `func NewBundleFromPartition(t *meta.Meta, job *model.Job, partition *model.PartitionInfo)`
2: tidy moved to constructors--- `NewBundleFromOptions`
3: the `newBundleFromPolicyOrDirectOptions` function is built